src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala

```diff
@@ -0,0 +1,197 @@
+import scala.concurrent.duration._
+import scala.concurrent.ExecutionContext.Implicits.global
+
+
+
+case class BalancesSnapshot(height: Height, balance: Long, assetBalances: Map[ByteArrayWrapper, Long])
+
+
+class ErgoWalletActor(seed: String,
+                      addressEncoder: ErgoAddressEncoder) extends Actor with ScorexLogging {
+
+  import ErgoWalletActor._
+  import Registry._
+
+  //todo: pass as parameter, add to config
+  val boxSelector: BoxSelector = DefaultBoxSelector
+
+  private val prover = new ErgoProvingInterpreter(seed)
+
+  private var height = 0
+  private var lastBlockUtxoRootHash = ADDigest @@ Array.fill(32)(0: Byte)
+
+  private val trackedAddresses: mutable.Buffer[ErgoAddress] =
+    mutable.Buffer(prover.dlogPubkeys: _ *).map(P2PKAddress.apply)
+
+  private val trackedBytes: mutable.Buffer[Array[Byte]] = trackedAddresses.map(addressEncoder.contentBytes)
+
+  //todo: make resolveUncertainty(boxId, witness)
+  private def resolveUncertainty(): Unit = {
+    Registry.nextUncertain().foreach { uncertainBox =>
+      val box = uncertainBox.box
+
+      val lastUtxoDigest = AvlTreeData(lastBlockUtxoRootHash, 32)
+
+      val testingTx = UnsignedErgoLikeTransaction(
+        IndexedSeq(new UnsignedInput(box.id)),
+        IndexedSeq(new ErgoBoxCandidate(1L, Values.TrueLeaf))
+      )
+
+      val context =
+        ErgoLikeContext(height + 1, lastUtxoDigest, IndexedSeq(box), testingTx, box, ContextExtension.empty)
+
+      prover.prove(box.proposition, context, testingTx.messageToSign) match {
+        case Success(_) =>
+          log.info(s"Uncertain box is mine! $uncertainBox")
+          Registry.makeTransition(uncertainBox, uncertainBox.makeCertain())
+        case Failure(_) =>
+        //todo: remove after some time? remove spent after some time?
+      }
+    }
+  }
+
+  def scan(tx: ErgoTransaction, heightOpt: Option[Height]): Boolean = {
+    tx.inputs.foreach { inp =>
+      val boxId = ByteArrayWrapper(inp.boxId)
+      if (Registry.registryContains(boxId)) {
+        Registry.makeTransition(boxId, ProcessSpending(tx, heightOpt))
+      }
+    }
+
+    tx.outputCandidates.zipWithIndex.exists { case (outCandidate, outIndex) =>
+      trackedBytes.find(t => outCandidate.propositionBytes.containsSlice(t)) match {
+        case Some(_) =>
+          val idxShort = outIndex.toShort
+          val box = outCandidate.toBox(tx.serializedId, idxShort)
+          val bu = heightOpt match {
+            case Some(h) => UnspentOnchainBox(tx, idxShort, h, box, Uncertain)
+            case None => UnspentOffchainBox(tx, idxShort, box, Uncertain)
+          }
+          bu.register()
+          true
+        case None =>
+          false
+      }
+    }
+  }
+
+  private def extractFromBlock(fb: ErgoFullBlock): Int = {
+    height = fb.header.height
+    lastBlockUtxoRootHash = fb.header.stateRoot
+    fb.transactions.count(tx => scan(tx, Some(height)))
+  }
+
+  override def receive: Receive = {
```

Comment:
cyclomatic complexity is reported here. I think it worth splitting the method to several handlers as we usually do